libvast/src/transform_steps/drop.cpp

```diff
@@ -0,0 +1,121 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2021 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include "vast/arrow_table_slice.hpp"
+#include "vast/concept/convertible/data.hpp"
+#include "vast/concept/convertible/to.hpp"
+#include "vast/error.hpp"
+#include "vast/logger.hpp"
+#include "vast/plugin.hpp"
+#include "vast/table_slice_builder_factory.hpp"
+#include "vast/transform.hpp"
+#include "vast/type.hpp"
+
+#include <arrow/type.h>
+
+namespace vast {
+
+namespace {
+
+/// The configuration of a project transform step.
+struct drop_step_configuration {
+  /// The key suffixes of the fields to drop.
+  std::vector<std::string> fields = {};
+
+  /// Support type inspection for easy parsing with convertible.
+  template <class Inspector>
+  friend auto inspect(Inspector& f, drop_step_configuration& x) {
+    return f(x.fields);
+  }
+
+  /// Enable parsing from a record via convertible.
+  static inline const record_type& layout() noexcept {
+    static auto result = record_type{
+      {"fields", list_type{string_type{}}},
+    };
+    return result;
+  }
+};
+
+/// Drops the specifed fields from the input.
+class drop_step : public transform_step {
+public:
+  explicit drop_step(drop_step_configuration config) noexcept
+    : config_{std::move(config)} {
+    // nop
+  }
+
+  caf::error
+  add(type layout, std::shared_ptr<arrow::RecordBatch> batch) override {
+    VAST_DEBUG("drop step adds batch");
+    // Apply the transformation.
+    auto transform_fn
+      = [&](struct record_type::field, std::shared_ptr<arrow::Array>) noexcept
+      -> std::vector<
+        std::pair<struct record_type::field, std::shared_ptr<arrow::Array>>> {
+      return {};
+    };
+    auto transformations = std::vector<indexed_transformation>{};
+    for (const auto& field : config_.fields)
+      for (auto&& index : caf::get<record_type>(layout).resolve_key_suffix(
+             field, layout.name()))
+        transformations.push_back({std::move(index), transform_fn});
+    std::sort(transformations.begin(), transformations.end());
```

Comment:
Not clear from the control flow why this needs sorting; maybe comment; not urgent though.

Reply:
It's a precondition of the `transform_columns` function necessarily.